presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PullUpExpressionInLambdaRules.java

```diff
@@ -0,0 +1,348 @@
+            if (specialForm.getForm().equals(SpecialFormExpression.Form.BIND)) {
+                return false;
+            }
+            Map<RowExpression, Boolean> validRowExpressionMap = specialForm.getArguments().stream().distinct().collect(toImmutableMap(identity(), x -> x.accept(this, context)));
+            if (context.equals(Boolean.TRUE)) {
+                boolean allArgumentsValid = validRowExpressionMap.values().stream().allMatch(x -> x.equals(Boolean.TRUE));
+                if (!allArgumentsValid) {
+                    candidates.addAll(validRowExpressionMap.entrySet().stream()
+                            .filter(x -> x.getValue().equals(Boolean.TRUE))
+                            .filter(x -> x.getKey() instanceof CallExpression || x.getKey() instanceof SpecialFormExpression)
+                            .map(Map.Entry::getKey)
+                            .collect(toImmutableList()));
+                }
+                return allArgumentsValid && determinismEvaluator.isDeterministic(specialForm);
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitLambda(LambdaDefinitionExpression lambda, Boolean context)
+        {
+            if (lambda.getBody().accept(this, true)) {
+                candidates.add(lambda.getBody());
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitVariableReference(VariableReferenceExpression reference, Boolean context)
+        {
+            return inputVariables.contains(reference);
+        }
+
+        @Override
+        public Boolean visitConstant(ConstantExpression literal, Boolean context)
+        {
+            return true;
+        }
+
+        @Override
+        public Boolean visitInputReference(InputReferenceExpression reference, Boolean context)
+        {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unexpected input reference in expression");
+        }
+    }
+
+    private static class ExpressionRewriter
+            implements RowExpressionVisitor<RowExpression, Void>
+    {
+        private final Map<RowExpression, VariableReferenceExpression> expressionMap;
+
+        public ExpressionRewriter(Map<RowExpression, VariableReferenceExpression> expressionMap)
+        {
+            this.expressionMap = ImmutableMap.copyOf(expressionMap);
+        }
+
+        @Override
+        public RowExpression visitCall(CallExpression call, Void context)
+        {
+            RowExpression rewritten = new CallExpression(
+                    call.getSourceLocation(),
+                    call.getDisplayName(),
+                    call.getFunctionHandle(),
+                    call.getType(),
+                    call.getArguments().stream().map(argument -> argument.accept(this, null)).collect(toImmutableList()));
+            if (expressionMap.containsKey(rewritten)) {
```

Comment:
Can you create new only when needed. ie., only if some arg is rewritten

Reply:
Done